policy-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/policy/kafka/KafkaContractTest.java

```diff
@@ -0,0 +1,359 @@
+import org.springframework.util.LinkedMultiValueMap;
+import org.testcontainers.containers.KafkaContainer;
+import scala.concurrent.duration.FiniteDuration;
+
+import uk.gov.gchq.palisade.contract.policy.ContractTestData;
+import uk.gov.gchq.palisade.service.policy.PolicyApplication;
+import uk.gov.gchq.palisade.service.policy.model.AuditErrorMessage;
+import uk.gov.gchq.palisade.service.policy.model.PolicyRequest;
+import uk.gov.gchq.palisade.service.policy.model.StreamMarker;
+import uk.gov.gchq.palisade.service.policy.model.Token;
+import uk.gov.gchq.palisade.service.policy.stream.ConsumerTopicConfiguration;
+import uk.gov.gchq.palisade.service.policy.stream.ProducerTopicConfiguration;
+import uk.gov.gchq.palisade.service.policy.stream.PropertiesConfigurer;
+import uk.gov.gchq.palisade.service.policy.stream.SerDesConfig;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toMap;
+import static org.apache.kafka.clients.admin.AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertAll;
+
+/**
+ * An external requirement of the service is to connect to a pair of kafka topics.
+ * The upstream "resource" topic is written to by the resource-service and read by this service.
+ * The downstream "rule" topic is written to by this service and read by the attribute-masking-service.
+ * Upon writing to the upstream topic, appropriate messages should be written to the downstream topic.
+ */
+@SpringBootTest(
+        classes = PolicyApplication.class,
+        webEnvironment = WebEnvironment.RANDOM_PORT,
+        properties = {"akka.discovery.config.services.kafka.from-config=false", "spring.cache.caffeine.spec=expireAfterWrite=1m, maximumSize=100"}
+)
+@Import({KafkaContractTest.KafkaInitializer.Config.class})
+@ContextConfiguration(initializers = {KafkaContractTest.KafkaInitializer.class})
+@ActiveProfiles({"caffeine", "akkatest", "pre-population"})
+class KafkaContractTest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaContractTest.class);
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    @Autowired
+    private TestRestTemplate restTemplate;
+    @Autowired
+    private ActorSystem akkaActorSystem;
+    @Autowired
+    private Materializer akkaMaterializer;
+    @Autowired
+    private ConsumerTopicConfiguration consumerTopicConfiguration;
+    @Autowired
+    private ProducerTopicConfiguration producerTopicConfiguration;
+
+    @Test
+    @DirtiesContext
+    void testVariousRequestSets() {
```

Comment:
Would appreciate documentation of each of the tests.  This is an example of a complex test and takes an effort to understand.

Reply:
added javadoc